File: metavision/sdk/cv3d/samples/metavision_model_3d_tracking/metavision_model_3d_tracking.cpp
/**********************************************************************************************************************
 * Copyright (c) Prophesee S.A. - All Rights Reserved                                                                 *
 *                                                                                                                    *
 * Subject to Prophesee Metavision Licensing Terms and Conditions ("License T&C's").                                  *
 * You may not use this file except in compliance with these License T&C's.                                           *
 * A copy of these License T&C's is located in the "licensing" folder accompanying this file.                         *
 **********************************************************************************************************************/

#include <opencv2/opencv.hpp>
#include <boost/program_options.hpp>
#include <boost/filesystem.hpp>
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/json_parser.hpp>

#include <metavision/sdk/driver/camera.h>
#include <metavision/sdk/core/algorithms/shared_cd_events_buffer_producer_algorithm.h>
#include <metavision/sdk/core/algorithms/periodic_frame_generation_algorithm.h>
#include <metavision/sdk/cv/algorithms/trail_filter_algorithm.h>
#include <metavision/sdk/cv/utils/camera_geometry.h>
#include <metavision/sdk/cv/utils/camera_geometry_factory.h>
#include <metavision/sdk/core/utils/mostrecent_timestamp_buffer.h>
#include <metavision/sdk/cv3d/algorithms/model_3d_tracking_algorithm.h>
#include <metavision/sdk/cv3d/algorithms/model_3d_detection_algorithm.h>
#include <metavision/sdk/cv3d/utils/model_3d_processing.h>
#include <metavision/sdk/ui/utils/event_loop.h>
#include <metavision/sdk/ui/utils/mt_window.h>

namespace po = boost::program_options;
namespace fs = boost::filesystem;
namespace pt = boost::property_tree;

struct Config {
    std::string base_path;
    std::string raw_path;
    std::string biases_path;
    std::string model_path;
    std::string model_pose_path;
    std::string calibration_path;
    std::uint32_t display_acc_time_us_;
    std::uint32_t n_detections_;
    std::uint32_t n_events_;
    Metavision::timestamp n_us_;
    Metavision::timestamp detection_period_us_;
    float display_fps_;
    bool no_display;
    bool realtime_playback_speed;
};

class Pipeline {
public:
    Pipeline(const Config &config) : config_(config) {
        is_tracking_ = false;
        n_detection_ = 0;

        /// [LOAD_3D_MODEL_BEGIN]
        if (!Metavision::load_model_3d_from_json(config.model_path, model_))
            throw std::runtime_error("Impossible to load the 3D model from " + config.model_path);
        /// [LOAD_3D_MODEL_END]

        if (config_.raw_path.empty()) {
            camera_ = Metavision::Camera::from_first_available();

            if (!config_.biases_path.empty())
                camera_.biases().set_from_file(config.biases_path);
        } else {
            camera_ = Metavision::Camera::from_file(
                config_.raw_path, Metavision::FileConfigHints().real_time_playback(config_.realtime_playback_speed));
        }

        const auto width  = camera_.geometry().width();
        const auto height = camera_.geometry().height();

        cam_geometry_ = Metavision::load_camera_geometry<float>(config.calibration_path);
        if (!cam_geometry_)
            throw std::runtime_error("Impossible to load the camera calibration from " + config.calibration_path);

        load_init_pose(config.model_pose_path);

        T_c_w_ = T_c_w_init_;

        time_surface_.create(height, width, 2);

        /// [INSTANTIATE_ALGOS_BEGIN]
        detection_algo_ =
            std::make_unique<Metavision::Model3dDetectionAlgorithm>(*cam_geometry_, model_, time_surface_);

        tracking_algo_ = std::make_unique<Metavision::Model3dTrackingAlgorithm>(*cam_geometry_, model_, time_surface_);
        /// [INSTANTIATE_ALGOS_END]

        frame_generation_algo_ = std::make_unique<Metavision::PeriodicFrameGenerationAlgorithm>(
            width, height, config.display_acc_time_us_, config.display_fps_);

        frame_generation_algo_->set_output_callback(
            std::bind(&Pipeline::frame_callback, this, std::placeholders::_1, std::placeholders::_2));

        /// [CD_PRODUCER_CALLBACK_BEGIN]
        Metavision::SharedEventsBufferProducerParameters params;
        auto cb   = [&](Metavision::timestamp ts, const Buffer &b) { buffers_.emplace(b); };
        producer_ = std::make_unique<Metavision::SharedCdEventsBufferProducerAlgorithm>(params, cb);
        /// [CD_PRODUCER_CALLBACK_END]

        set_detection_params();

        camera_.cd().add_callback(
            std::bind(&Pipeline::cd_processing_callback, this, std::placeholders::_1, std::placeholders::_2));

        if (!config_.no_display) {
            window_ = std::make_unique<Metavision::MTWindow>("3D Model tracking", width, height,
                                                             Metavision::BaseWindow::RenderMode::BGR);
            window_->set_keyboard_callback(std::bind(&Pipeline::ui_key_callback, this, std::placeholders::_1,
                                                     std::placeholders::_2, std::placeholders::_3,
                                                     std::placeholders::_4));
        }
    }

    void run() {
        camera_.start();

        while (camera_.is_running()) {
            if (!window_) {
                std::this_thread::yield();
                continue;
            }

            if (window_->should_close())
                break;

            Metavision::EventLoop::poll_and_dispatch(10); // no need to rush
        }

        camera_.stop();
    }

private:
    void load_init_pose(const std::string &path) {
        pt::ptree root;

        pt::read_json(path, root);

        const auto &cam_node = root.get_child("camera_pose");

        Eigen::Matrix4f T_w_c;

        int row = 0;
        for (const auto &row_node : cam_node.get_child("T_w_c")) {
            int col = 0;
            for (const auto &col_node : row_node.second) {
                T_w_c(row, col) = col_node.second.get_value<float>();
                ++col;
            }
            ++row;
        }

        T_c_w_init_ = T_w_c.inverse();
    }

    void process_buffers_queue() {
        // As the async_algorithm class is not reentrant (i.e. calling flush inside a process_async call would
        // create internal conflicts) buffers are not processed directly when created (i.e. inside the callback
        // passed to the producer) but added to a queue and processed afterwards instead.
        // After processing the buffer, we check whether the tracking status has changed. If so, we change the
        // parameters of the events buffers producer accordingly. As a consequence a flush is called in the
        // async_algorithm which causes the creation of a new buffer that can safely be added to the queue and
        // processed.
        /// [BUFFERS_QUEUE_PROCESSING_LOOP_BEGIN]
        while (!buffers_.empty()) {
            const bool prev_is_tracking = is_tracking_;

            const auto buffer = buffers_.front();
            buffers_.pop();

            const auto begin = buffer->cbegin();
            const auto end   = buffer->cend();

            if (is_tracking_) {
                is_tracking_ = tracking_algo_->process_events(begin, end, T_c_w_);
            } else if (detection_algo_->process_events(begin, end, T_c_w_, &visible_edges_, &detected_edges_)) {
                // we wait for several detections before considering the model as detected to avoid false positive
                // detections
                is_tracking_ = (++n_detection_ > config_.n_detections_);
            }

            // the frame generation algorithm processing can trigger a call to show_async which can trigger a reset of
            // the tracking if the space bar has been pressed.
            frame_generation_algo_->process_events(begin, end);

            if (prev_is_tracking != is_tracking_) {
                if (is_tracking_)
                    set_tracking_params(std::prev(buffer->cend())->t); // triggers the creation of a new buffer
                else
                    set_detection_params(); // triggers the creation of a new buffer
            }
        }
        /// [BUFFERS_QUEUE_PROCESSING_LOOP_END]
    }

    void set_detection_params() {
        MV_LOG_INFO() << "set detection param";
        T_c_w_ = T_c_w_init_;
        detection_algo_->set_init_pose(T_c_w_);
        producer_->set_processing_n_us(config_.detection_period_us_);
    }

    void set_tracking_params(Metavision::timestamp ts) {
        MV_LOG_INFO() << "set tracking param";
        tracking_algo_->set_previous_camera_pose(ts, T_c_w_);
        n_detection_ = 0;
        producer_->set_processing_mixed(config_.n_events_, config_.n_us_);
    }

    /// [FRAME_GENERATOR_CALLBACK_BEGIN]
    void frame_callback(Metavision::timestamp ts, cv::Mat &frame) {
        cv::putText(frame, std::to_string(ts), cv::Point(0, 10), cv::FONT_HERSHEY_DUPLEX, 0.5,
                    (is_tracking_ ? cv::Scalar(0, 255, 0) : cv::Scalar(0, 0, 255)));

        // Be careful, here the events and the 3D model are not rendered in a tightly synchronized way, meaning that
        // some shifts might occur. However, most of the time they should not be noticeable
        if (is_tracking_) {
            Metavision::select_visible_edges(T_c_w_, model_, visible_edges_);
            Metavision::draw_edges(*cam_geometry_, T_c_w_, model_, visible_edges_, frame, cv::Scalar(0, 255, 0));
            cv::putText(frame, "tracking", cv::Point(0, 30), cv::FONT_HERSHEY_DUPLEX, 0.5, cv::Scalar(0, 255, 0));
        } else {
            Metavision::draw_edges(*cam_geometry_, T_c_w_, model_, visible_edges_, frame, cv::Scalar(0, 0, 255));
            Metavision::draw_edges(*cam_geometry_, T_c_w_, model_, detected_edges_, frame, cv::Scalar(0, 255, 0));
            cv::putText(frame, "detecting", cv::Point(0, 30), cv::FONT_HERSHEY_DUPLEX, 0.5, cv::Scalar(0, 0, 255));
        }

        if (window_)
            window_->show_async(frame);
    }
    /// [FRAME_GENERATOR_CALLBACK_END]

    /// [CD_CAMERA_CALLBACK_BEGIN]
    void cd_processing_callback(const Metavision::EventCD *begin, const Metavision::EventCD *end) {
        producer_->process_events(begin, end);

        process_buffers_queue();
    }
    /// [CD_CAMERA_CALLBACK_END]

    void ui_key_callback(Metavision::UIKeyEvent key, int scancode, Metavision::UIAction action, int mods) {
        if (action == Metavision::UIAction::RELEASE) {
            if (key == Metavision::UIKeyEvent::KEY_SPACE)
                is_tracking_ = false;
            else if (key == Metavision::UIKeyEvent::KEY_Q || key == Metavision::UIKeyEvent::KEY_ESCAPE)
                window_->set_close_flag();
        }
    }

    using Buffer = Metavision::SharedCdEventsBufferProducerAlgorithm::SharedEventsBuffer;

    const Config config_;
    Metavision::Model3d model_;
    Metavision::Camera camera_;
    Metavision::MostRecentTimestampBufferT<Metavision::timestamp> time_surface_;
    std::unique_ptr<Metavision::CameraGeometryBase<float>> cam_geometry_;
    std::unique_ptr<Metavision::Model3dDetectionAlgorithm> detection_algo_;
    std::unique_ptr<Metavision::Model3dTrackingAlgorithm> tracking_algo_;
    std::unique_ptr<Metavision::PeriodicFrameGenerationAlgorithm> frame_generation_algo_;
    std::unique_ptr<Metavision::MTWindow> window_;

    std::unique_ptr<Metavision::SharedCdEventsBufferProducerAlgorithm> producer_;
    std::queue<Buffer> buffers_;

    volatile bool is_tracking_;
    std::uint32_t n_detection_;
    Eigen::Matrix4f T_c_w_init_;
    Eigen::Matrix4f T_c_w_;
    std::set<size_t> visible_edges_, detected_edges_;
};

bool parse_command_line(int argc, char *argv[], Config &config) {
    const std::string program_desc("3D model detection and tracking\n");

    po::options_description options_desc;
    po::options_description base_options("Base options");
    // clang-format off
    base_options.add_options()
        ("help,h", "Produce help message.")
        ("input-base-path,i", po::value<std::string>(&config.base_path), "Base path used to load a recording, a 3D model, a camera calibration and an initialization pose. Apart from the recording, every file can also be set individually")
        ("input-raw-file,r", po::value<std::string>(&config.raw_path), "Path to a RAW file. Ignored if a base path is set")
        ("input-bias-file,b", po::value<std::string>(&config.biases_path), "Path to a BIAS file. Ignored if a path to a RAW file is set")
        ("input-model-file,m", po::value<std::string>(&config.model_path), "Path to a JSON file containing the description of a 3D model")
        ("input-pose-file,p", po::value<std::string>(&config.model_pose_path), "Path to a JSON file containing a camera pose used to detect the 3D model")
        ("input-calibration-file,c", po::value<std::string>(&config.calibration_path), "Path to a JSON file containing the camera's calibration")
        ;
    // clang-format on

    po::options_description detection_options("Detection options");
    // clang-format off
    detection_options.add_options()
        ("num-detections", po::value<std::uint32_t>(&config.n_detections_)->default_value(10), "Number of successive valid detection to consider the model as detected")
        ("detection-period", po::value<Metavision::timestamp>(&config.detection_period_us_)->default_value(10000), "Amount of time after which a detection is attempted")
        ;
    // clang-format on

    po::options_description tracking_options("Tracking options");
    // clang-format off
    tracking_options.add_options()
        ("n-events", po::value<std::uint32_t>(&config.n_events_)->default_value(5000), "Number of events after which a tracking step is attempted")
        ("n-us", po::value<Metavision::timestamp>(&config.n_us_)->default_value(10000), "Amount of time after which a tracking step is attempted")
        ;
    // clang-format on

    po::options_description display_options("Display options");
    // clang-format off
    display_options.add_options()
        ("disp-accumulation-time,a", po::value<std::uint32_t>(&config.display_acc_time_us_)->default_value(5000), "Accumulation time in us used to generate frames for display")
        ("fps,f", po::value<float>(&config.display_fps_)->default_value(30.f), "Display's fps")
        ("no-display,d", po::bool_switch(&config.no_display)->default_value(false), "Disable output display window")
        ("realtime-playback-speed", po::value<bool>(&config.realtime_playback_speed)->default_value(true), "Replay events at speed of recording if true, otherwise as fast as possible")
        ;
    // clang-format on

    options_desc.add(base_options);
    options_desc.add(detection_options);
    options_desc.add(tracking_options);
    options_desc.add(display_options);

    po::variables_map vm;
    try {
        po::store(po::command_line_parser(argc, argv).options(options_desc).run(), vm);
        po::notify(vm);
    } catch (po::error &e) {
        MV_LOG_ERROR() << program_desc;
        MV_LOG_ERROR() << options_desc;
        MV_LOG_ERROR() << "Parsing error:" << e.what();
        return false;
    }

    if (vm.count("help")) {
        MV_LOG_INFO() << program_desc;
        MV_LOG_INFO() << options_desc;
        return false;
    }

    fs::path root_path;
    fs::path base_path;
    if (config.base_path.empty()) {
        if (config.raw_path.empty()) {
            MV_LOG_ERROR() << "You should provide either a base path or a path to a RAW file";
            return false;
        }

        const fs::path raw_path(config.raw_path);
        if (!fs::exists(raw_path)) {
            MV_LOG_ERROR() << "Invalid RAW path:" << raw_path.string();
            return false;
        }

        root_path = raw_path.parent_path();
        base_path = root_path / raw_path.stem();
    } else {
        base_path = fs::path(config.base_path);
        root_path = base_path.parent_path();

        const std::string extension = base_path.extension().string();
        if (extension != "") {
            MV_LOG_ERROR() << "Invalid base path. Remove the extension" << extension;
            if (extension == ".raw")
                MV_LOG_ERROR() << "Or use -r instead of -i to run the app on a RAW file.";
            return false;
        }

        if (!fs::is_directory(root_path)) {
            MV_LOG_ERROR() << "Invalid base path:" << base_path.string();
            return false;
        }

        if (!config.raw_path.empty())
            config.raw_path.clear();
    }

    if (config.model_path.empty())
        config.model_path = base_path.string() + ".json";

    if (!fs::exists(config.model_path)) {
        MV_LOG_ERROR() << "The path to the model file is invalid:" << config.model_path;
        return false;
    }

    if (config.model_pose_path.empty())
        config.model_pose_path = base_path.string() + "_init_pose.json";

    if (!fs::exists(config.model_pose_path)) {
        MV_LOG_ERROR() << "The path to the model's pose file is invalid:" << config.model_pose_path;
        return false;
    }

    if (config.calibration_path.empty())
        config.calibration_path = (root_path / fs::path("calibration.json")).string();

    if (!fs::exists(config.calibration_path)) {
        MV_LOG_ERROR() << "The path to the camera's calibration file is invalid:" << config.calibration_path;
        return false;
    }

    MV_LOG_INFO() << "root path:" << root_path;
    MV_LOG_INFO() << "base path:" << config.base_path;
    MV_LOG_INFO() << "raw path:" << config.raw_path;
    MV_LOG_INFO() << "biases path:" << config.biases_path;
    MV_LOG_INFO() << "model path:" << config.model_path;
    MV_LOG_INFO() << "model pose path:" << config.model_pose_path;
    MV_LOG_INFO() << "calibration path:" << config.calibration_path;

    return true;
}

int main(int argc, char *argv[]) {
    Config config;
    if (!parse_command_line(argc, argv, config))
        return 1;

    try {
        Pipeline p(config);

        p.run();
    } catch (const std::runtime_error &e) {
        MV_LOG_ERROR() << e.what();
        return 1;
    }

    return 0;
}

File: metavision/sdk/cv3d/python_samples/metavision_model_3d_tracking/metavision_model_3d_tracking.py
# Copyright (c) Prophesee S.A. - All Rights Reserved
#
# Subject to Prophesee Metavision Licensing Terms and Conditions ("License T&C's").
# You may not use this file except in compliance with these License T&C's.
# A copy of these License T&C's is located in the "licensing" folder accompanying this file.

"""
Script to detect and track an object, the 3D model of which is provided by the user.
"""

import cv2
import numpy as np
import os
import json

from metavision_core.event_io import RawReader
from metavision_sdk_core import MostRecentTimestampBuffer, PeriodicFrameGenerationAlgorithm
from metavision_sdk_cv import load_camera_geometry
from metavision_sdk_cv3d import EigenMatrix4f, Model3dDetectionAlgorithm, Model3dTrackingAlgorithm
import metavision_sdk_cv3d as mv_cv3d  # Static methods: load_model_3d_from_json, select_visible_edges, draw_edges
from metavision_sdk_ui import EventLoop, BaseWindow, MTWindow, UIAction, UIKeyEvent


def parse_args():
    import argparse
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(description='Model 3D Tracking sample.',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    # Base options
    base_options = parser.add_argument_group('Base options')
    base_options.add_argument('-i', '--input-base-path', dest='base_path', default="",
                              help="Base path used to load a recording, a 3D model, a camera calibration and an "
                              "initialization pose. Apart from the recording, every file can also be set individually.")
    base_options.add_argument('-r', '--input-raw-file', dest='raw_path', default="",
                              help="Path to input RAW file. Ignored if a base path is set.")
    base_options.add_argument('-m', '--input-model-file', dest='model_path', default="",
                              help="Path to a JSON file containing the description of a 3D model.")
    base_options.add_argument('-p', '--input-pose-file', dest='model_pose_path', default="",
                              help="Path to a JSON file containing a camera pose used to detect the 3D model.")
    base_options.add_argument('-c', '--input-calibration-file', dest='calibration_path', default="",
                              help="Path to a JSON file containing the camera's calibration.")
    base_options.add_argument('--process-from', dest='process_from', type=int, default=0,
                              help='Start time to process events (in us).')
    base_options.add_argument('--process-to', dest='process_to', type=int, default=None,
                              help='End time to process events (in us).')

    # Detection options
    detection_options = parser.add_argument_group('Detection options')
    detection_options.add_argument(
        '--num-detections', dest='n_detections', type=int, default=10,
        help='Number of successive valid detection to consider the model as detected.')
    detection_options.add_argument('--detection-period', dest='detection_period_us', type=int,
                                   default=10000, help='Amount of time after which a detection is attempted.')

    # Tracking options
    tracking_options = parser.add_argument_group('Tracking options')
    tracking_options.add_argument('--n-events', dest='n_events', type=int, default=5000,
                                  help='Number of events after which a tracking step is attempted.')
    tracking_options.add_argument('--n-us', dest='n_us', type=int, default=10000,
                                  help='Amount of time after which a tracking step is attempted.')

    # Outcome Options
    outcome_options = parser.add_argument_group('Outcome options')
    outcome_options.add_argument('-a', '--accumulation-time', dest='display_acc_time_us', type=int, default=5000,
                                 help='Accumulation time in us used to generate frames for display.')
    outcome_options.add_argument('-f', '--fps', dest='display_fps', type=float, default=30.,
                                 help="Display's fps.")
    outcome_options.add_argument(
        '-o', '--out-video', dest='out_video', type=str, default="",
        help="Path to output AVI file to save slow motion video. A frame is generated after each process of the algorithm. The video "
        "will be written only for processed events.")

    args = parser.parse_args()

    if args.process_to and args.process_from > args.process_to:
        print(f"The processing time interval is not valid. [{args.process_from,}, {args.process_to}]")
        exit(1)

    if not args.base_path:
        if not args.raw_path:
            print("You should provide either a base path or a path to a RAW file.")
            exit(2)
        if not os.path.exists(args.raw_path):
            print("Invalid RAW path: " + args.raw_path)
            exit(3)
        args.base_path = os.path.splitext(args.raw_path)[0]  # Remove the .raw extension
    else:
        extension = os.path.splitext(args.base_path)[1]
        if extension != "":
            print("Invalid base path. Remove the extension " + extension)
            if extension == ".raw":
                print("Or use -r instead of -i to run the app on a RAW file.")
            exit(4)

        root_path = os.path.dirname(args.base_path)  # Directory containing the base path
        if not os.path.exists(root_path):
            print("Invalid base path: " + args.base_path)
            exit(5)
        args.raw_path = ""  # Ignored if a base path is set

    # Model
    if not args.model_path:
        args.model_path = args.base_path + ".json"
    if not os.path.exists(args.model_path):
        print("Invalid model path: " + args.model_path)
        exit(6)

    # Model pose
    if not args.model_pose_path:
        args.model_pose_path = args.base_path + "_init_pose.json"
    if not os.path.exists(args.model_pose_path):
        print("Invalid model's pose path: " + args.model_pose_path)
        exit(7)

    # Calibration
    if not args.calibration_path:
        args.calibration_path = os.path.join(os.path.dirname(args.base_path), "calibration.json")
    if not os.path.exists(args.calibration_path):
        print("Invalid camera calibration path: " + args.calibration_path)
        exit(8)

    print("Base path       :" + args.base_path)
    print("Raw path        :" + args.raw_path)
    print("Model path      :" + args.model_path)
    print("Model pose path :" + args.model_pose_path)
    print("Calibration path:" + args.calibration_path)

    return args


def load_init_pose(model_pose_path):
    with open(model_pose_path) as json_file:
        data = json.load(json_file)["camera_pose"]
        inv_pose = data["T_w_c"]
        pose = np.linalg.inv(inv_pose)

        T_c_w = EigenMatrix4f()
        T_c_w.numpy()[...] = pose[...]

        return T_c_w


def main():
    """ Main """
    args = parse_args()

    # [LOAD_3D_MODEL_BEGIN]
    # Load 3D Model
    model_3d = mv_cv3d.load_model_3d_from_json(args.model_path)
    if not model_3d:
        print("Impossible to load the 3D model from " + args.model_path)
        exit(9)
    # [LOAD_3D_MODEL_END]

    T_c_w_init = load_init_pose(args.model_pose_path)  # Load initial pose
    T_c_w = T_c_w_init.copy()

    # Raw Reader on Camera or RAW file
    reader = RawReader(args.raw_path)

    height, width = reader.get_size()  # Camera Geometry
    camera_geometry = load_camera_geometry(args.calibration_path)
    if not camera_geometry:
        print("Impossible to load the camera calibration from " + args.calibration_path)
        exit(10)

    # Time Surface
    time_surface = MostRecentTimestampBuffer(height, width, 2)

    # [INSTANTIATE_ALGOS_BEGIN]
    # Detection and Tracking algorithms
    detection_algo = Model3dDetectionAlgorithm(camera_geometry, model_3d, time_surface)
    tracking_algo = Model3dTrackingAlgorithm(camera_geometry, model_3d, time_surface)
    # [INSTANTIATE_ALGOS_END]

    visible_edges = set()
    detected_edges = set()
    is_tracking = False
    n_detection = 0

    def set_detection_params():
        nonlocal T_c_w
        T_c_w = T_c_w_init.copy()
        detection_algo.set_init_pose(T_c_w_init)

    def set_tracking_params(ts):
        nonlocal n_detection
        tracking_algo.set_previous_camera_pose(ts, T_c_w)
        n_detection = 0

    set_detection_params()

    # Window - Graphical User Interface (Display tracking results and process keyboard events)
    with MTWindow(title="Model 3D Tracking", width=width, height=height, mode=BaseWindow.RenderMode.BGR) as window:
        if args.out_video:
            fourcc = cv2.VideoWriter_fourcc('M', 'J', 'P', 'G')
            video_name = args.out_video + ".avi"
            video_writer = cv2.VideoWriter(video_name, fourcc, 20, (width, height))

        def keyboard_cb(key, scancode, action, mods):
            nonlocal is_tracking
            if action != UIAction.RELEASE:
                return
            if key == UIKeyEvent.KEY_ESCAPE or key == UIKeyEvent.KEY_Q:
                window.set_close_flag()
            elif key == UIKeyEvent.KEY_SPACE:
                is_tracking = False

        window.set_keyboard_callback(keyboard_cb)
        print("Press 'q' to leave the program.\n"
              "Press 'Space' to reset the tracking.\n")

        # [FRAME_GENERATOR_CALLBACK_BEGIN]
        # Periodic Event Frame Generator
        def periodic_frame_gen_cb(ts, cv_frame):
            nonlocal is_tracking
            nonlocal visible_edges
            nonlocal detected_edges

            cv2.putText(cv_frame, str(ts), (0, 10), cv2.FONT_HERSHEY_DUPLEX, 0.5,
                        (0, 255, 0) if is_tracking else (0, 0, 255))
            # Be careful, here the events and the 3D model are not rendered in a tightly synchronized way,
            # meaning that some shifts might occur. However, most of the time they should not be noticeable
            if is_tracking:
                visible_edges = mv_cv3d.select_visible_edges(T_c_w, model_3d)
                mv_cv3d.draw_edges(camera_geometry, T_c_w, model_3d, visible_edges, cv_frame, (0, 255, 0))
                cv2.putText(cv_frame, "tracking", (0, 30), cv2.FONT_HERSHEY_DUPLEX, 0.5, (0, 255, 0))
            else:
                mv_cv3d.draw_edges(camera_geometry, T_c_w, model_3d, visible_edges, cv_frame, (0, 0, 255))
                mv_cv3d.draw_edges(camera_geometry, T_c_w, model_3d, detected_edges, cv_frame, (0, 255, 0))
                cv2.putText(cv_frame, "detecting", (0, 30), cv2.FONT_HERSHEY_DUPLEX, 0.5, (0, 0, 255))
            window.show_async(cv_frame)
            if args.out_video:
                video_writer.write(cv_frame)
        # [FRAME_GENERATOR_CALLBACK_END]

        events_frame_gen_algo = PeriodicFrameGenerationAlgorithm(width, height, args.display_acc_time_us,
                                                                 args.display_fps)
        events_frame_gen_algo.set_output_callback(periodic_frame_gen_cb)

        # [PROCESSING_LOOP_BEGIN]
        # Process events
        reader.seek_time(args.process_from)
        while (not reader.is_done()) and (args.process_to is None or reader.current_time < args.process_to):
            # Dispatch system events to the window
            EventLoop.poll_and_dispatch()

            prev_is_tracking = is_tracking

            if is_tracking:
                evs = reader.load_n_events(args.n_events)
                is_tracking = tracking_algo.process_events(evs, T_c_w)
            else:
                evs = reader.load_delta_t(args.detection_period_us)
                success, visible_edges, detected_edges = detection_algo.process_events(evs, T_c_w)
                if success:
                    # We wait for several detections before considering the model as detected
                    # to avoid false positive detections
                    n_detection += 1
                    is_tracking = (n_detection > args.n_detections)

            # The frame generation algorithm processing can trigger a call to show which can trigger
            # a reset of the tracking if the space bar has been pressed.
            events_frame_gen_algo.process_events(evs)

            if prev_is_tracking != is_tracking:
                if is_tracking:
                    set_tracking_params(evs["t"][-1])
                else:
                    set_detection_params()
        # [PROCESSING_LOOP_END]

            if window.should_close():
                break

        if args.out_video:
            video_writer.release()
            print("Video has been saved in " + video_name)


if __name__ == "__main__":
    main()
